decode-audio-promise/index.html

```diff
@@ -0,0 +1,130 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width" />
+
+    <title>Web Audio API examples: decodeAudioData()</title>
+  </head>
+
+  <body>
+    <h1>Web Audio API examples: decodeAudioData()</h1>
+
+    <button class="play">Play</button>
+    <button class="stop">Stop</button>
+
+    <h2>Set playback rate</h2>
+    <input
+      class="playback-rate-control"
+      type="range"
+      min="0.25"
+      max="3"
+      step="0.05"
+      value="1"
+    />
+    <span class="playback-rate-value">1.0</span>
+
+    <h2>Set loop start and loop end</h2>
+    <input
+      class="loopstart-control"
+      type="range"
+      min="0"
+      max="20"
+      step="1"
+      value="0"
+    />
+    <span class="loopstart-value">0</span>
+
+    <input
+      class="loopend-control"
+      type="range"
+      min="0"
+      max="20"
+      step="1"
+      value="0"
+    />
+    <span class="loopend-value">0</span>
+  </body>
+  <script>
+    window.addEventListener("DOMContentLoaded", loadPage, false);
+
+    // define variables
+    const audioCtx = new AudioContext();
+    let buffer;
+    let source;
+
+    const play = document.querySelector(".play");
+    const stop = document.querySelector(".stop");
+
+    const playbackControl = document.querySelector(".playback-rate-control");
+    const playbackValue = document.querySelector(".playback-rate-value");
+    playbackControl.setAttribute("disabled", "disabled");
+
+    const loopstartControl = document.querySelector(".loopstart-control");
+    const loopstartValue = document.querySelector(".loopstart-value");
+    loopstartControl.setAttribute("disabled", "disabled");
+
+    const loopendControl = document.querySelector(".loopend-control");
+    const loopendValue = document.querySelector(".loopend-value");
+    loopendControl.setAttribute("disabled", "disabled");
+
+    function loadPage() {
+      fetchAudio("viper").then((buf) => { // executes when buffer has been decoded
+          let max = Math.floor(buf.duration); // in this case buf === global buffer
+          loopstartControl.setAttribute("max", max);
+          loopendControl.setAttribute("max", max);
+      });
+    }
+
+    // use fetch() to load an audio track, and
+    // decodeAudioData to decode it and stick it in global buffer variable.
+    // We put the buffer into the source in play.onclick()
+    async function fetchAudio(name) {
+      try {
+        let rsvp = await fetch(`${name}.mp3`);
+        buffer   = await audioCtx.decodeAudioData(await rsvp.arrayBuffer());
+        return buffer; // returns a Promise, buffer is arg for .then((arg) => {})
```

Comment:
```suggestion
        return audioCtx.decodeAudioData(await rsvp.arrayBuffer());
```

(just a suggestion. But I don't think you need to await `decodeAudioData` here.)

Reply:
From what I tested, await is appropriate for decodeAudioData().  I like it the way it is because I think the comment is helpful and because of await.

Reply:
Sorry, I hadn't realised you were setting the global `buffer` there, so you do need the `await`.

Reply:
Yea, it's that way because of precisely what the comment explains.  The return value isn't really the return value because it's all async.

Reply:
I'm not sure if we are talking at cross purposes, but it's not the return value of `fetchAudio` that's the problem - that's going to return a promise that resolves to the buffer anyway, so there's no effective difference AFAIK between:

```js
const buffer = await audioCtx.decodeAudioData(...);
return buffer;
```

...and just

```js
return audioCtx.decodeAudioData(...);
```

The issue is that you're assigning the result of `audioCtx.decodeAudioData(...)` to the global `buffer` variable, so we need that to be the actual buffer, not a promise. For example, if you were doing that in the caller, you could just return the result of `decodeAudioData` directly:

```js
    function loadPage() {
      fetchAudio("viper").then((buf) => {
        buffer = buf;
        ... 
      });
    }

    // use fetch() to load an audio track, and
    // decodeAudioData to decode it and stick it in global buffer variable.
    // We put the buffer into the source in play.onclick()
    async function fetchAudio(name) {
      try {
        const rsvp = await fetch(`${name}.mp3`);
        return audioCtx.decodeAudioData(await rsvp.arrayBuffer());
      } catch (err) {
        console.error(`Unable to fetch the audio file: ${name}
error: ${err.message}`);
      }
    }
```

...but it's fine the way it is.

It's not ideal because there is a potential race where we click "Play" before the promise has resolved, and `buffer` is still undefined. But it's not worth worrying about.

Reply:
> It's not ideal because there is a potential race where we click "Play" before the promise has resolved, and buffer is still undefined. But it's not worth worrying about.

The issue about the play button is part of this goofy disabling/enabling scheme that my first pass "fixed", combined with the update to only fetch the buffer once.  I can fix that here by disabling the button inside the button tag and enabling it in the `.then()'.  Or I can do what the original example does and wait until the user clicks play to download the file.  Let me know if you'd like me to make one of those changes.
  
Yes, `buffer` could certainly be assigned in the `.then()`.  Six of one...  It doesn't affect the play button issue you raise.

Reply:
> I can fix that here by disabling the button inside the button tag and enabling it in the `.then()`. 

I think this is probably worth doing.

Reply:
Done.